### DISCUSSION TITLE
Handle a dioProvider when access token changes

### DISCUSSION THREAD

**Discussion by mrskme** [rrousselGit/riverpod#1818] [Q&A]
I have a userProvider, dioProvider and a credentialsProvider. The home page watches the userProvider which watches dioProvider which watches credentialsProvider. When the user logs out the credentials is set to null, the dioProvider rebuilds which in turn makes the userProvider rebuild and it makes the api call again but without authentication. What's the way to do this in riverpod without making unnecessary calls?

```dart
final userProvider = FutureProvider<User>((ref) async {
  final dio = ref.watch(dioProvider);
  final response = await dio.get('auth/me');
  final User user = User.fromJson(response.data);
  return user;
});

final dioProvider = Provider<Dio>((ref) {
  final credentials = ref.watch(credentialsProvider);
  final headers = {
    HttpHeaders.authorizationHeader:
        '${credentials!.tokenType} ${credentials.accessToken}',
  };
  final options = BaseOptions(
    baseUrl: dotenv.env['API_URL']!,
    headers: headers,
  );
  Dio dio = Dio(options);
  return dio;
});
```

**Comment by rrousselGit** [comment]
The behavior you described sounds like a good thing IMO.

But you could use ref.listen instead of ref.watch:


```dart

final dioProvider = Provider<Dio>((ref) {
  final dio = Dio();
  ref.listen<Credential?>(credentialsProvider, (prev, credentials) {
    dio.options = BaseOptions(...);
  }, fireImmediately: true);
  return dio;
});
```

Since the `Dio` instance won't change when `credentialsProvider` updates, that won't re-trigger your network request.
